nestkernel/sonata_connector.cpp

```diff
@@ -0,0 +1,667 @@
+
+      close_dsets_();
+      reset_params_();
+
+    } // end iteration over population groups
+
+    // Close H5 objects in scope
+    edges_grp->close();
+    file->close();
+
+  } // end iteration over edge files
+}
+
+
+H5::H5File*
+SonataConnector::open_file_( std::string& fname )
+{
+  H5::H5File* file = nullptr;
+  try
+  {
+    file = new H5::H5File( fname, H5F_ACC_RDONLY );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open HDF5 file " + fname + ": " + e.getDetailMsg() );
+  }
+  return file;
+}
+
+H5::Group*
+SonataConnector::open_group_( const H5::H5File* file, const std::string& grp_name )
+{
+  H5::Group* group = nullptr;
+  try
+  {
+    group = new H5::Group( file->openGroup( grp_name ) );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open HDF5 group " + grp_name + " in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+  return group;
+}
+
+H5::Group*
+SonataConnector::open_group_( const H5::Group* group, const std::string& grp_name )
+{
+  H5::Group* subgroup = nullptr;
+  try
+  {
+    subgroup = new H5::Group( group->openGroup( grp_name ) );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open HDF5 group " + grp_name + " in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+  return subgroup;
+}
+
+
+void
+SonataConnector::open_required_dsets_( const H5::Group* pop_grp )
+{
+  try
+  {
+    src_node_id_dset_ = pop_grp->openDataSet( "source_node_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open source_node_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  try
+  {
+    tgt_node_id_dset_ = pop_grp->openDataSet( "target_node_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open target_node_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  try
+  {
+    edge_type_id_dset_ = pop_grp->openDataSet( "edge_type_id" );
+  }
+  catch ( const H5::Exception& e )
+  {
+    throw KernelException( "Could not open edge_type_id dataset in " + cur_fname_ + ": " + e.getDetailMsg() );
+  }
+
+  // Consistency checks
+  const auto num_tgt_node_ids = get_nrows_( tgt_node_id_dset_ );
+
+  // Ensure that target and source population have the same size
+  if ( num_tgt_node_ids != get_nrows_( src_node_id_dset_ ) )
+  {
+    throw KernelException(
+      "target_node_id and source_node_id datasets in " + cur_fname_ + " must be of the same size" );
+  }
+
+  // Ensure that edge_type_id dataset size is consistent with the number of target node ids
+  if ( num_tgt_node_ids != get_nrows_( edge_type_id_dset_ ) )
+  {
+    throw KernelException( "target_node_id and edge_type_id datasets in " + cur_fname_ + " must be of the same size" );
+  }
+}
+
+void
+SonataConnector::try_open_edge_group_id_dsets_( const H5::Group* edge_id_grp )
+{
+  // TODO: Currently only works if the edge file has a single edge id group
```

Comment:
Could we assert() this here?

Reply:
We currently verify that only one edge group is present before this function is called:

https://github.com/nicolossus/nest-simulator/blob/689502d40e6cfc9b8fd4bcdace83db1d5b3bbbef/nestkernel/sonata_connector.cpp#L128-L132

In its current implementation, the function only receives a pointer to a single edge group. I'm therefore not sure if we should also do an assert here. We would then need to pass the number of found edge groups to the function, which already should have raised an error if there are more than one.

The comment is more directed towards that we need to rework this function when we add support for more than
one edge group.